Issue 5328 - Scroll effects update: custom zones & units

Fybex opened the pull request:
# Description

<!---
Please include a summary of the changes and the related issue.
Please also include relevant motivation and context.
--->

Fixes #5328 

Changelog:
- custom zones (no more start, mid, end zones, add and set zone in any percentage)
![image](https://github.com/maxi-blocks/maxi-blocks/assets/46112160/e15c10f8-1298-4f26-8836-ea528e67f8a9)
- units for vertical, horizontal and blur
![image](https://github.com/maxi-blocks/maxi-blocks/assets/46112160/e134b57b-5cc9-45cd-8653-63ebb9b228f4)
- fixed responsive status disable (possible to disable scroll effects on some breakpoint)
- migrator (to not break blocks with scroll effects from previous versions)

### Also, a new dependency was added: `react-slider`. So don't forget to `npm ci --legacy-peer-deps`

<!---
Please describe the tests that you ran to verify your changes.
Provide instructions so we can reproduce.
Please also list any relevant details for your test configuration.
--->

# Test checklist

<!--- Please remove the unnecessary checkbox --->

**_ Front/Back Testing _**

-   [ ] Test custom zones
-   [ ] Test units
-   [ ] Test shortcuts
-   [ ] Test migrator (save some block with scroll effects on master and switch to this branch)
-   [ ] Check that the settings work on frontend

**_ Pre-Code Testing _**

-   [ ] Test custom zones
-   [ ] Test units
-   [ ] Test shortcuts
-   [ ] Test migrator (save some block with scroll effects on master and switch to this branch)
-   [ ] Check that the settings work on frontend
-   [ ] Check no commented code and no unnecessary imports
-   [ ] Standards of the project have been followed
-   [ ] No errors/warnings on console

# Checklist

-   [X] My code follows the style guidelines of this project
-   [X] I have performed a self-review of my code
-   [X] I have commented my code, particularly in hard-to-understand areas
-   [X] My changes generate no new warnings/errors


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced new scroll effects properties and enhanced control over scroll-related transformations.
  - Added new UI components for managing scroll effects with improved user interaction.
  - Expanded scroll effects settings to include a wider range of attributes and units.

- **Enhancements**
  - Enhanced the scroll effects logic to support additional breakpoints and scroll data handling.
  - Improved the slider component's styling in the editor for better visual appeal and usability.

- **Refactor**
  - Refactored scroll effects handling, including state management and event handling in the UI components.
  - Updated scroll effects attribute generation to be more flexible and support new types.

- **Bug Fixes**
  - Fixed issues with scroll effect attribute migration to ensure smooth transitions between versions.

- **Documentation**
  - Documented new effect properties and their settings for clarity in use and configuration.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Fybex committed:
Add `react-slider`

Fybex committed:
Raw version of scroll effects custom zones

Fybex committed:
Scroll effects zones adaptation

Fybex committed:
Fix scroll effects responsive

Fybex committed:
Remove unnecessary double attr for zones

kyrapieterse committed:
Scroll Kyra tweaks

🏂

kyrapieterse committed:
Update editor.scss

🎩

Fybex committed:
Merge branch 'issue-5328' of https://github.com/maxi-blocks/maxi-blocks into issue-5328

Fybex committed:
Migrate scroll shortcuts

Fybex committed:
Scroll effects migrator

Fybex committed:
Hide `data-scroll-effect-type` if it's empty

Fybex committed:
Clean classnames for scroll effect slider

Fybex committed:
ScrollEffectUniqueControl: clean + fix reset

Fybex committed:
Use exported `scrollTypes`

Fybex committed:
Scroll effects: add different units support

to 'vertical', 'horizontal', 'blur' scroll types

Fybex committed:
Fix scroll effects `getCurrentBreakpoint`

Fybex committed:
Make possible to have only one zone

Fybex committed:
Use previous value on addition, select on deletion

Fybex committed:
Switch unit to px back on shortcut change

Fybex committed:
Change selected thumb on deletion if it's selected

Fybex committed:
Fix migrator for vertical scroll effects

Fybex committed:
Merge branch 'master' into issue-5328

Fybex committed:
Update unit tests

Olekrut reviewed (changes requested):
Wowoowowowo) It's so amazing 🥇 Well done 🎉

Few points:
1) IMPORTANT!
![image](https://github.com/maxi-blocks/maxi-blocks/assets/88893350/fffed52c-0c92-43a3-915f-b0dc445607fe)
 This 100% is height of target block, but it should be height of screen. About viewport entry, guess this point in %, actually, should work as it. Because it  points 2 and maybe 3 don't work correct. But I am not sure how it should work with elements which have height 100% or more or a bit less....
Ping me in pm if you need explanation.
2) rotate and scale don't not work correct ( important ) 
https://jam.dev/c/9a1459c7-5260-4893-a090-653e4e606282
3) vertical scroll does not work ( important )
https://jam.dev/c/30714050-8a76-4ed5-80d8-dd87b00ddf7f
4) Adding button "+" is appearing only if difference between points more than 20%
https://jam.dev/c/6dd5f95e-1cf5-4c07-b4a4-eaee99d8e4fd
5) hard to pick overlapped point - if difference 1-3% ( is not very important )
https://jam.dev/c/42b2079e-71f3-4312-9203-f9627b73b8ae
6) when we are removing point > position field is empty and working but for one of points > guess, if we're removing point should be picked one of existed ( next or previous point ). Check after cases when we are removing first and last points. ( is not very important )
https://jam.dev/c/baa92ad6-7f72-48a5-ae79-05243d950210
7) UI/UX conflict with Scroll combinations and direction presets. 
https://jam.dev/c/dff2c40c-a4de-4966-866f-6c797bbe4042
https://jam.dev/c/be9c3114-1e7d-43af-8850-d9b8dfc99f5b
8) Adding button (+) is moving higher and lower than should
https://jam.dev/c/da9483c7-06fc-40d9-b57d-e612306131b5

Fybex committed:
Merge branch 'master' into issue-5328

Fybex committed:
Delete vertical scroll old attrs on migration

Fybex committed:
Merge branch 'master' into issue-5328

Fybex committed:
Add screen zone option

Fybex committed:
`forEach` -> `for`: block zone reference logic

Fybex committed:
Fix block zone logic, simplify `scrollTransform`

Fybex committed:
Improve show add button condition

Fybex committed:
Fix active thumb switch on thumb deletion

Fybex committed:
Merge branch 'master' into issue-5328

Fybex committed:
Hide add button if it's not inside of slider

Fybex committed:
Fix old `getMaxiBlockAttributes` scroll support

Fybex commented:
1,2,3 points: added screen zone option as default
![image](https://github.com/maxi-blocks/maxi-blocks/assets/46112160/71604530-3f6c-4bbf-9650-90acfe7056c9)
added this switch to remain old functionality

4, 6, 7, 8 points: fixed
5 point: low priority, going to derive later all issues like this to separate issue 👍 

Scroll effects have limited responsive for the moment (only enable/disable option). Do you think we need more advanced support at this point, @Olekrut?

Olekrut reviewed (changes requested):
1) need to add scale X and  Y; rotate X, Y, Z; like separated options, but we should have ability to use them in one time

2) impossible to click to some of the elements now ( remove point impossible sometimes, need to pick point for it ) - is not important
https://jam.dev/c/2af4195f-b3fa-40ab-933d-8f4374f12134
3) yes, would be good to change values on responsive - but is not very important for now can be realized in next issue
4) Would be good in the future to have ability to add scroll effects to BG elements

Fybex committed:
Fix speed in rotate shortcuts

Fybex committed:
Scroll `rotateX`, `rotateY`, `scaleX`, `scaleY`

Fybex committed:
Fix add button hide if it overlaps thumb

Fybex committed:
Abstract migrator part of `getMaxiBlockAttributes`

Fybex committed:
Is block zone to true for migrated blocks

Fybex committed:
Set to `source.fixAll.eslint` auto valid value

Fybex committed:
Fix shortcut

Fybex commented:
> 1. need to add scale X and  Y; rotate X, Y, Z; like separated options, but we should have ability to use them in one time
Done
![image](https://github.com/maxi-blocks/maxi-blocks/assets/46112160/2f9e3ed6-46f6-4aaa-9a6b-be15e87dd860)
![image](https://github.com/maxi-blocks/maxi-blocks/assets/46112160/d31774c5-d6c2-4043-bcdf-0c798ca4c006)


> 2. impossible to click to some of the elements now ( remove point impossible sometimes, need to pick point for it ) - is not important
>    https://jam.dev/c/2af4195f-b3fa-40ab-933d-8f4374f12134

Thanks, fixed

> 3. yes, would be good to change values on responsive - but is not very important for now can be realized in next issue

#5432 👍 

> 4. Would be good in the future to have ability to add scroll effects to BG elements

#5433  👍

Fybex committed:
Update unit tests

Olekrut commented:
When we have ability to enable scale, rotate with their parts X and Y! 
It is ok for now, but in the future need to fix it in some ways
1) When we are using some scale or rotate, another one should be unavailable
or
2) to make it possible, to use all part of scroll effect in one time

Fybex committed:
Fix using multiple scale, rotate effects

Fybex commented:
> When we have ability to enable scale, rotate with their parts X and Y! It is ok for now, but in the future need to fix it in some ways
> 
> 1. When we are using some scale or rotate, another ordinate should be unavailable
>    or
> 2. to make it possible, to use all part of scroll effect in one time

Fixed it, now rotate, rotateX, rotateY can work simultaneously (same with scale)

> 
> @Fybex OOO, and I ve found for what we can use your idea , so to add it to separate issue pls https://jam.dev/c/1411db02-af6d-4d65-a82e-2adb5e805472 [full page zone.txt](https://github.com/maxi-blocks/maxi-blocks/files/14249524/full.page.zone.txt) would be good if we will have ability to make full page or container as scroll zone

Thanks, #5437 👍

Olekrut reviewed (approved):
1) UI. Should we rename it to RotateZ? 
![image](https://github.com/maxi-blocks/maxi-blocks/assets/88893350/b18e6e86-53d6-4fa5-833d-9552c4a8ede9)
2) It is not important, but  when we are reloading/opening page with scroll styles, styles are applying to blocks only in few seconds after 
https://jam.dev/c/d5b30484-7c49-4ad7-b5ea-3045f84b3f38

If it is long, you can move these points to next issue

elzadj commented on .vscode/settings.json:
```diff
@@ -1,7 +1,7 @@
 	"editor.formatOnSave": true,
 	"editor.codeActionsOnSave": {
-		"source.fixAll.eslint": true
+		"source.fixAll.eslint": "explicit"
```
VSCode true is set to "explicit" since VSCode 1.85.0. It seems boolean values were supported until 1.84.0.

References

https://code.visualstudio.com/updates/v1_85#_code-actions-on-save-and-auto https://code.visualstudio.com/updates/v1_83#_code-actions-on-save-and-auto-save
